### ISSUE TITLE
Support ReadWriteMany

### ISSUE THREAD

**Issue by kallisti5** [digitalocean/csi-digitalocean#136]
### What did you do? (required. The issue will be **closed** when not provided.)
Allow a volume to be accessed by multiple pods.

### What did you expect to happen?
A volume could be accessed by multiple pods.

### Configuration (**MUST** fill this out):
```
111s        Warning   ProvisioningFailed     PersistentVolumeClaim   failed to provision volume with StorageClass "do-block-storage": rpc error: code = InvalidArgument desc = invalid volume capabilities requested. Only SINGLE_NODE_WRITER is supported ('accessModes.ReadWriteOnce' on Kubernetes)
```

* manifests, such as pvc, deployments, etc.. you used to reproduce:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: git-deployment
  namespace: default
  labels:
    app: git
spec:
  replicas: 2
  selector:
    matchLabels:
      app: git
  template:
    metadata:
      labels:
        app: git
    spec:
      containers:
      - name: gerrit
        image: haiku/gerrit:2.16
        volumeMounts:
        - name: gerrit-data
          mountPath: /var/gerrit/review_site
        env:
          - name: AUTH_TYPE
            value: "OAUTH"
          - name: GITWEB_TYPE
            value: cgit
          - name: JAVA_HEAPLIMIT
            value: "1024m"
          - name: SMTP_SERVER
            value: "smtp"
          - name: SMTP_SERVER_PORT
            value: "25"
          - name: SMTP_FROM
            value: "Gerrit <review@review.haiku-os.org>"
          - name: SMTP_CONNECT_TIMEOUT
            value: "30"
      - name: cgit
        image: haiku/cgit
        env:
          - name: CGIT_TITLE
            value: "Haiku's repositories"
          - name: CGIT_DESC
            value: "Presented by cgit on kubernetes"
          - name: CGIT_VROOT
            value: "/"
          - name: CGIT_SECTION_FROM_STARTPATH
            value: "1"
        volumeMounts:
        - name: gerrit-data
          mountPath: /var/git
      volumes:
      - name: gerrit-data
        persistentVolumeClaim:
          claimName: gerrit-data-pvc
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: gerrit-data-pvc
spec:
  accessModes:
  - ReadWriteMany
  #- ReadWriteOnce
  resources:
    requests:
      storage: 6Gi
  storageClassName: do-block-storage-persist
---
apiVersion: v1
kind: Service
metadata:
  name: cgit
  namespace: default
spec:
  selector:
    app: git
  ports:
  - name: web
    port: 80
    targetPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: review
  namespace: default
spec:
  selector:
    app: git
  ports:
  - name: web
    port: 80
    targetPort: 8080
  - name: git
    port: 2222
    targetPort: 29418
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: git
  namespace: default
spec:
  rules:
  - host: cgit.haiku-os.org
    http:
      paths:
      - path: /
        backend:
          serviceName: cgit
          servicePort: web
  - host: review.haiku-os.org
    http:
      paths:
      - path: /
        backend:
          serviceName: review 
          servicePort: web
```

Storage (just for persistant storage per 
```
kind: StorageClass
metadata:
  annotations:
    storageclass.kubernetes.io/is-default-class: "false"
    name: "do-block-storage-persist"
    provisioner: "dobs.csi.digitalocean.com"
  name: do-block-storage-persist
provisioner: dobs.csi.digitalocean.com
reclaimPolicy: Retain
volumeBindingMode: Immediate
```

**Comment by sagebind** [comment]
This isn't going to ever work I imagine, because the plugin is still just using plain old [block storage volumes](https://www.digitalocean.com/products/block-storage/) under the hood, and _those_ do not support mounting to more than one droplet. So unless that changes, then there's nothing this plugin can do to support `ReadWriteMany`.

**Comment by fatih** [comment]
Yeap, DO Block Storage doesn't support ReadWriteMany unfortunately. Closing this as there is not much we can do here. Thanks for the feedback nevertheless.

**Comment by JCMais** [comment]
Is using a nfs-server the best option here in this case?

**Comment by adamwg** [comment]
@JCMais Yes, if you need ReadWriteMany storage in a DOKS cluster, you are best to deploy an NFS server. You could do this outside the cluster (on a droplet), or use something like the [openebs dynamic NFS provisioner](https://github.com/openebs/dynamic-nfs-provisioner) to manage it inside the cluster.

**Comment by kallisti5** [comment]
gcp / gke offers managed nfs.   "It would be nice" if DO offered something similar for ReadWriteMany

**Comment by TripShade** [comment]
> @JCMais Yes, if you need ReadWriteMany storage in a DOKS cluster, you are best to deploy an NFS server. You could do this outside the cluster (on a droplet), or use something like the [openebs dynamic NFS provisioner](https://github.com/openebs/dynamic-nfs-provisioner) to manage it inside the cluster.

Im having issues with the helm charts of this project use https://openebs.github.io/charts/nfs-operator.yaml it does work for me.

**Comment by matteius** [comment]
I've run into this issue today -- I have a saleor docker container with a web and celery worker.  I am trying to create a volume between them that is `ReadWriteMany` because an image is uploaded to the web container and a celery worker gets the file path to process the file upload and with the default mode it cannot find the file (separate containers).  I just realized from this issue that the way I am trying to configure it with `ReadWriteMany` won't be supported, but now I am at a loss as to what the solution to my issue for sharing the file access between two docker containers in Kubernetes would be.

**Comment by timoreimann** [comment]
@matteius are you able to schedule the two containers / pods onto the same worker node (via, say, node affinity rules)? If so, then they would be able to access the same volume.

If that's not doable, then an alternative like NFS provisioner (as mentioned in this thread) or some shared file system sitting on top of DO block storage would come to mind.

**Comment by matteius** [comment]
Thanks @timoreimann  -- I tried first this library but that really wasn't going as planned and then someone else pointed out to me that saleor uses dango-storages behind the scenes so I configured it to upload directly to S3 instead of to the local file system to get a work around, instead of trying to use some kind of shared block storage between containers.  My understanding is you are correct that NFS would work too, but the level of setup and complexity for that is much higher.  Thanks!

**Comment by timoreimann** [comment]
Agreed that setting up NFS or a shared file-system comes with a certain effort. Co-scheduling dependent pods would seem slightly easier, but if you've got it sorted out by "sharing" through S3 (AFAIU it) then that's another reasonable option.

**Comment by kallisti5** [comment]
Digital Ocean just needs a reliable + cost effective managed NFS (or other RWX provider)
https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes


With that said, you *CAN* make multiple things use RWO.  You just have to explicitly schedule them on the same node pool node.   I did this for Haiku here:
https://github.com/haiku/infrastructure/blob/master/deployments/rsync.yml#L20

That affinity rule keeps the deployment on the same node pool node as the other deployment consuming the PVC. It's messy as hell though.

**Comment by kallisti5** [comment]
> At least in my experience, it is almost always better to re-architect your application to not require sharing file systems between multiple nodes except for a few cases where you are OK with the slow performance of NFS.

No doubts here.  The reality is however legacy non-microservice focused applications almost always need RWX in some capacity.   Either the application is  designed around complex webs of multiple services accessing persistent data, *or* you have a single pod which needs RWX to accomplish zero-downtime updates. (since you can't use the RollingUpdate deployment strategy with RWO)

lack of cheap RWX is always a pain point migrating existing infrastructure with any sizable PV claims to k8s.

**Comment by joeldeteves** [comment]
Why did this issue get closed? This is a much-needed feature.

**Comment by TripShade** [comment]
> Why did this issue get closed? This is a much-needed feature.

I agree, right now the only way I get a usable Kubernetes is by deploying openebs-nfs-provisioner and using their storage class as rwx. This should be working out of the box.

**Comment by droslean** [comment]
Is this feature going to be supported? Why this issue is closed?

**Comment by kallisti5** [comment]
I feel like this tldr is it's not possible to attach a single block device to multiple physical kubernetes nodes, so this will never work in the current design.

DigitalOcean needs to work on a managed (and cost effective) NFS/GlusterFS/Ceph storage provider.  That would be an "easy" path to RWX

**Comment by droslean** [comment]
@kallisti5 Thanks for the response. So that means that DO is not even working on that, and if they are, it doesn't sound like it will be ready soon enough. So we have to figure this out our own way. Maybe we should deploy our own NFS controller or whatever. That's really bad because Kubernetes is already managed by DO. The easiest way to move forward is to just change the cloud provider instead. AWS and GCS already support RWX.

**Comment by matteius** [comment]
@droslean I ended up just using digital ocean spaces and I forget the reason I wanted this in the first place.

**Comment by droslean** [comment]
> @droslean I ended up just using digital ocean spaces and I forget the reason I wanted this in the first place.

Can you elaborate more? How you are using spaces as volumes?

**Comment by matteius** [comment]
> How you are using spaces as volumes?

I am not, I reengineered my solution to not use volumes per-se and to instead use boto3 to upload/retrieve things to Spaces.

**Comment by kallisti5** [comment]
tldr; pretty much anything that requires on-disk storage should be converted to indirect bucket / s3 storage if possible.

If your app or use-case absolutely requires on-disk storage, you need to scale your cluster so an individual node can accept n + 1  workloads (the ability to run an extra pod of something). Then ensure your deployment pods can only run on one physical node.

That way, when you do upgrades / rollouts you can have one pod start up, one pod shutdown and rotate through updates on a single physical node.

In the configuration above, RWO will work fine.  It's limiting, but functional.  You do loose node redundancy though since that node going down will result in a brief outage as workloads start on another.

**Comment by droslean** [comment]
It nice workaround but it won't work in my case. There is no way to control which node the pod will run since the DO auto scaler doesn't even support labeling the nodes or there is no way to add a node selector to the deployment since the name of the nodes always changing. I could turn the auto scaler off and try to hardcode everything but this is even worse. Unfortunately, my apps need disk storage and I can't change it. 

I think the best solution is to go with something like https://github.com/openebs/dynamic-nfs-provisioner but I am not sure how the costs will be affected. All options lead to leaving DO and starting using AWS or GCP.

**Comment by timoreimann** [comment]
@droslean FWIW, DOKS supports persistent node pool labels, meaning you can associate a label with all current and future nodes of a pool. When cluster-autoscaler brings up new nodes of a pool, the labels will be there right from the start.

Similarly, there's support for associating persistent node pool taints.

See the [corresponding API reference section](https://docs.digitalocean.com/reference/api/api-reference/#operation/kubernetes_add_nodePool) for details (fields `labels` and `taints`).

**Comment by droslean** [comment]
> @droslean FWIW, DOKS supports persistent node pool labels, meaning you can associate a label with all current and future nodes of a pool. When cluster-autoscaler brings up new nodes of a pool, the labels will be there right from the start.
> 
> Similarly, there's support for associating persistent node pool taints. AFAIU, the combination of these two paired with affinity rules should give you control over where workloads should land even when autoscaling is involved.
> 
> See the [corresponding API reference section](https://docs.digitalocean.com/reference/api/api-reference/#operation/kubernetes_add_nodePool) for details (fields `labels` and `taints`).

Thanks @timoreimann! So this is only available through the API directly or I can control the labels via doctl or the GUI?

**Comment by timoreimann** [comment]
@droslean it is supported by the API, godo, and doctl.

The GUI / cloud control panel does not support it (yet).

**Comment by Qasim-Aziz** [comment]
facing same problem i need RWX but with csi-s3 driver at least  i am able to use digitalocean spaces as volume under my statefullset.

**Comment by ryanovas** [comment]
It's a shame that in 2025 we don't have a solution for this, sadly there are many open source tools that need a volume and we can't just rewrite them to use S3. DO might be the only cloud provider I'm aware at this point that doesn't have some kind of RWX solution in their kubernetes offering...

**Comment by kubax2000** [comment]
This is now possible using DigitalOcean Network File Storage. An NFS share can be mounted in a DOKS cluster with the ReadWriteMany access mode.

Note that Network File Storage is currently available only in the NYC2, AMS3, ATL1, RIC1, and MKC1 regions. 

[How to Use NFS Storage with Kubernetes Clusters](https://docs.digitalocean.com/products/kubernetes/how-to/use-nfs-storage/)
